### PR TITLE
architecture.rst: update Ostro OS architecture 

### PR DESCRIPTION
**Pull request by mythi**
A very early draft to improve the architecture doc.

I'm not getting the diagram shown. The URL is 
https://github.com/mythi/meta-ostro/blob/arch-doc/doc/architecture/images/ostro_os_architecture.svg

What else would you like to read?

[skip ci]

### CODE REVIEW

**Comment by mythi** [comment]
@dbkinder @gvancuts

File: doc/architecture/architecture.rst [hunk 1]
```diff
@@ -1,16 +1,190 @@
...
+used in Ostro OS are commonly used in open source and any Ostro OS
+custom implementation is avoided. This has a big benefit because it
+means there is no extra porting/platform adaptation needed to get
+existing applications or services run in Ostro OS.
+
+2. Ostro OS is pre-compiled and configured: Ostro OS is based
+on the `Yocto Project_` and is provided
+as pre-compiled and configured distribution for IoT. There is no
+end-user compilation needed when starting from scratch with Ostro OS
+since the build system is configured to use the shared. In addition,
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: You're probably missing something after 'shared'?

File: doc/architecture/architecture.rst [hunk 2]
```diff
@@ -1,16 +1,190 @@
...
+Ostro |trade| OS Characteristics
+=================================
+
+In Ostro |trade| OS, there are four key characteristics that
+drive the architecture design:
+
+1. Ostro OS uses standard libraries and services: The building blocks
+used in Ostro OS are commonly used in open source and any Ostro OS
+custom implementation is avoided. This has a big benefit because it
+means there is no extra porting/platform adaptation needed to get
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: Would it be appropriate to change that to 'very little' as there would still be a little porting needed, if nothing else in the form of a app manifest to be created.

File: doc/architecture/architecture.rst [hunk 3]
```diff
@@ -1,16 +1,190 @@
...
+pre-compiled reference images are available for download.
+
+3. Ostro OS is secured and built with security in mind: Security is
+critical for IoT and it is also a focus in Ostro OS. The security
+design is to provide scalable options how security can be configured
+in Ostro based products. The security architecture is explained in
+more details in :ref: `system-and-security-architecture`.
+
+4. Ostro OS is easy to customize: While Ostro OS comes pre-compiled
+(binaries and images), it is important that the OS can be customized
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: I would emphasize that the pre-compiled images are *not* to be used in production.

File: doc/architecture/architecture.rst [hunk 4]
```diff
@@ -1,16 +1,190 @@
...
+the links provided there give a good start to Ostro OS customization.
+
+Ostro |trade| OS Architecture Stack
+===================================
+
+One way to descibe Ostro OS architecture is to build a stack from
+hardware platform up to an IoT Application layer. This stack is
+illustrated in Fig 1.
+
+.. image::images/ostro_os_architecture.png
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: Is the problem you're facing not simply due to a mismatch in the file extension? (s/png/svg)

File: doc/architecture/architecture.rst [hunk 5]
```diff
@@ -1,16 +1,190 @@
...
+to make it quick and easy to write software for IoT devices. The
+applications can we written using a flow-based programing (FBP) or
+more traditionally as C applications linking to Soletta's C based
+platform APIs.
+
+* `Node.JS`_ is a popular JavaScript runtime to
+run (IoT) web applications. Ostro OS provides the Node.JS runtime
+and selected JavaScript APIs (as Node.JS modules) to build IoT
+applications. The set of APIs include, e.g., a Javascript API for
+the OIC specification. The Javascript API set is kept aligned with
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: s/OIC specification/OCF specifications

File: doc/architecture/architecture.rst (hunk 4 above)
Comment (gvancuts) [inline doc/architecture/architecture.rst]: Can we update the picture to reflect the appropriate capitalization of some of these, e.g. ConnMan, BlueZ, IoTivity

Comment (dbkinder) [inline doc/architecture/architecture.rst]: wrong filename and missing a space after the ::
There are other .rst issues that i'm going to address and submit a PR to @mythi 's repo directly

File: doc/architecture/architecture.rst [hunk 6]
```diff
@@ -1,16 +1,190 @@
...
+3. Ostro OS is secured and built with security in mind: Security is
+critical for IoT and it is also a focus in Ostro OS. The security
+design is to provide scalable options how security can be configured
+in Ostro based products. The security architecture is explained in
+more details in :ref: `system-and-security-architecture`.
+
+4. Ostro OS is easy to customize: While Ostro OS comes pre-compiled
+(binaries and images), it is important that the OS can be customized
+to meet the diverse product needs in IoT space. With the help of
+Yocto Project build system used by Ostro OS, it is easy to customize
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: make the Yocto Project a link? (`Yocto Project`_)

File: doc/architecture/architecture.rst [hunk 7]
```diff
@@ -1,16 +1,190 @@
...
+as pre-compiled and configured distribution for IoT. There is no
+end-user compilation needed when starting from scratch with Ostro OS
+since the build system is configured to use the shared. In addition,
+pre-compiled reference images are available for download.
+
+3. Ostro OS is secured and built with security in mind: Security is
+critical for IoT and it is also a focus in Ostro OS. The security
+design is to provide scalable options how security can be configured
+in Ostro based products. The security architecture is explained in
+more details in :ref: `system-and-security-architecture`.
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: You have an extra space after :ref:

File: doc/architecture/architecture.rst [hunk 8]
```diff
@@ -1,16 +1,190 @@
...
+* In addition to Soletta and Node.JS, common programming languages/
+runtimes in IoT space are supported. Ostro OS provides Python runtime
+and Java Platform (OpenJDK8) that are popular in the IoT space.
+
+Frameworks
+----------
+
+Frameworks provide an abstraction to platform services. For making
+application deployment easy, the Ostro Application
+Framework :ref:`application_framework` provides tooling for
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: s/application_framework/application-framework

File: doc/architecture/architecture.rst [hunk 9]
```diff
@@ -1,16 +1,190 @@
...
+parses the application manifest files to systemd .service.
+
+`IoTivity`_ TODO
+
+Services
+--------
+
+The responsibility of system services is to bring up the system,
+manage wireless connections, and process' IPC. The components
+used are commonly used in all open source OSes: systemd, connman,
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: s/connman/ConnMan
s/bluez/BlueZ

File: doc/architecture/architecture.rst [hunk 10]
```diff
@@ -1,16 +1,190 @@
...
+bluez, and D-Bus.
+
+In addition to systemd and connectivity, Ostro OS comes with
+:ref: `software-update` technology that helps to support software
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: Extra space after :ref:

File: doc/architecture/architecture.rst [hunk 11]
```diff
@@ -1,16 +1,190 @@
...
+
+
+Linux Kernel and Hardware Adaptation
+-------------------------------------
+
+The hardware BSPs for Ostro OS run the Linux Kernel. The kernel
+provides the necessary drivers and hardware adaptation.
+
+Sensors and connectivity are critical for IoT devices. Therefore,
+Ostro OS maintains a dedicated page: :ref: `hardware` that describes
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: extra space after :ref:

File: doc/architecture/architecture.rst [hunk 12]
```diff
@@ -1,16 +1,190 @@
...
+how various peripherals can be run with :ref:`quick_start/platforms`
```
Comment (gvancuts) [inline doc/architecture/architecture.rst]: Out of curiosity (question to @dbkinder ), do we need to have the "quick_start/" prefix here or would :ref:`platforms` work equally well?

**Comment by gvancuts** [comment]
Hi @mythi ,

Thanks for kicking this off! I have made some comments (and also sent a PR to your branch adressing some of them). I have some more generic comments here:

1. The four pillars for Ostro are: Security, Connectivity, Sensors and Device Management. Can we reflect that better in the document? As an example, where you talk about services and frameworks, we should talk about the device update mechanism (at least mention it).

2. I have mixed feelings about having 'pre-compiled and pre-configured' and 'fully customizable' as two main characteristics. Both are true but I think we need to put them both in a single context and also explain that while Ostro comes pre-compiled and pre-configured, it's *not* meant to be deployed as is (it's mentionned in the customizable section but I think we need to make it more obvious).

Thanks

File: doc/architecture/architecture.rst (hunk 8 above)
Comment (dbkinder) [inline doc/architecture/architecture.rst]: I went through the document fixing broken :ref: name  syntax and getting the linking and formatting right and did a PR directly to Mikko’s repo rather than inundating this PR with lots of little fixes.
-- david

Comment (mythi) [inline doc/architecture/architecture.rst]: On Mon, Apr 25, 2016 at 6:59 PM, David Kinder <notifications@github.com>
wrote:

> In doc/architecture/architecture.rst
> <https://github.com/ostroproject/meta-ostro/pull/134#discussion_r60938910>
> :
>
> > +run (IoT) web applications. Ostro OS provides the Node.JS runtime
> > +and selected JavaScript APIs (as Node.JS modules) to build IoT
> > +applications. The set of APIs include, e.g., a Javascript API for
> > +the OIC specification. The Javascript API set is kept aligned with
> > +the API set provided by Soletta.
> > +
> > +* In addition to Soletta and Node.JS, common programming languages/
> > +runtimes in IoT space are supported. Ostro OS provides Python runtime
> > +and Java Platform (OpenJDK8) that are popular in the IoT space.
> > +
> > +Frameworks
> > +----------
> > +
> > +Frameworks provide an abstraction to platform services. For making
> > +application deployment easy, the Ostro Application
> > +Framework :ref:`application_framework` provides tooling for
>
> I went through the document fixing broken :ref: name syntax and getting
> the linking and formatting right and did a PR directly to Mikko’s repo
> rather than inundating this PR with lots of little fixes.
>
>
@dbkinder your changes are updated in this PR too. Thanks! I'll do the same
with Geoffroy's improvements shortly.

**Comment by mythi** [comment]
@gvancuts @dbkinder I've pushed our edits in this PR and added some more content. Also, the diagram should be updated.

File: doc/architecture/architecture.rst [hunk 13]
```diff
@@ -1,24 +1,234 @@
...
+#. **Ostro OS uses standard libraries and services**: The building blocks
+   used in Ostro OS are commonly used in many open-source projects and custom
+   implementations are avoided to the extent possible. This has a big benefit
+   as there is little porting or platform adaptation needed to get
+   existing applications or services to run on Ostro OS; often times, all that's
+   needed is to create a manifest for your application.
+
+#. **Ostro OS is pre-compiled and pre-configured**: Ostro OS leverages the
+   `Yocto Project`_ build system tools and is provided as a pre-compiled and pre-configured
+   distribution specifically tuned for IoT devices. Generating and compiling your own
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: ... specifically tuned for IoT device development.

[Do we need to lightly mention here that this pre-configured stuff is for development only?]

File: doc/architecture/architecture.rst [hunk 14]
```diff
@@ -1,24 +1,234 @@
...
+application deployment easier. The Ostro Application
+Framework :ref:`application-framework` provides tooling for
+application writers to get the applications running in an Ostro OS
+based image.
+
+The Ostro Application framework implements a `systemd.generator'_ that
+parses the application manifest files to generate a systemd service files.
+
+Ostro OS also pre-integrates `IoTivity`_ framework. IoTivity is an open source
+project that implements the full OCF specification by _`Open Connectivity Foundation`. In
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: move the underscore after the `name` for an outgoing link:  should be `Open Connectivity Foundation`_

File: doc/architecture/architecture.rst [hunk 15]
```diff
@@ -1,24 +1,234 @@
...
+--------
+
+The responsibility of system services is to bring the system up,
+manage connectivity and set up process' inter-process communication (IPC).
+The components Ostro OS use are commonly found in many open-source OSes: systemd,
+ConnMan, BlueZ, D-Bus and others.
+
+In addition to systemd and connectivity, Ostro OS comes with
+:ref:`software-update` technology that helps support software
+updates to deployed devices. That is also used to replace traditional
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: s/That/This technology/

File: doc/architecture/architecture.rst [hunk 16]
```diff
@@ -1,24 +1,234 @@
...
+package management. This means Ostro OS comes with no RPMs/DEBs. However,
+it can still be enabled if such feature is needed in a product/device.
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: s/it/RPMs or DEBs support/

File: doc/architecture/architecture.rst [hunk 17]
```diff
@@ -1,24 +1,234 @@
...
+
+Ostro OS 1.0 focuses on headless use-cases and thus no graphics and
+multimedia are enabled by default.
+
+Base Libraries
+--------------
+
+Standard Linux base libraries are used in Ostro. This means libraries
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: ... are used in Ostro; the same libraries

File: doc/architecture/architecture.rst [hunk 18]
```diff
@@ -1,24 +1,234 @@
...
+possible to easily extend the content of Ostro OS if there's a need to add new libraries.
+
+Linux Kernel and Hardware Adaptation
+-------------------------------------
+
+The hardware board support packages (BSP) for Ostro OS run the Linux Kernel. The kernel
+provides the necessary drivers and hardware adaptation.
+
+Sensors and connectivity are critical for IoT devices. Ostro OS reference Linux
+kernel configuration focuses on making good amount of sensors and connectivity peripherals
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: s/good amount/a good number/

**Comment by mythi** [comment]
update: rebased, updated based on feedback, force pushed

File: doc/architecture/architecture.rst [hunk 19]
```diff
@@ -1,24 +1,252 @@
...
+
+The following documents give more technical details about specific Ostro
+OS components and their usage.
 
 .. toctree::
    :maxdepth: 1
 
-   disk-layout
-   efi-boot
    software-update
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: @mythi  can you add     software-update-server     after this line and resubmit (this will get the updated doc that's already been merged into the table-of-contents   @gvancuts

**Comment by mythi** [comment]
@dbkinder @gvancuts your feedback/changes updated

**Comment by gvancuts** [comment]
:+1:

**Comment by dbkinder** [comment]
+1 thanks!

**Comment by mythi** [comment]
@barbieri have you seen this PR?

File: doc/architecture/architecture.rst [hunk 20]
```diff
@@ -1,24 +1,254 @@
...
+Framework :ref:`application-framework` provides tooling for
+application writers to get the applications running in an Ostro OS
+based image.
+
+The Ostro Application framework implements a `systemd.generator`_ that
+parses the application manifest file to automatically generate a systemd service file.
+
+Ostro OS also pre-integrates `IoTivity`_ framework. IoTivity is an open-source
+project that implements the full OCF specification by `Open Connectivity Foundation`_. In
+addition to `IoTivity`_, another implementation of the specification API is also
```
Comment (barbieri) [inline doc/architecture/architecture.rst]: don't go deep into details here, rephrase as "This connectivity framework is also available from Soletta and Node.js".

shouldn't matter Soletta is a different implementation, there is an ongoing talk to make IoTivity use Soletta in the core, nothing decided yet, but it may happen.

File: doc/architecture/architecture.rst [hunk 21]
```diff
@@ -1,24 +1,254 @@
...
+The following documents give more technical details about specific Ostro
+OS components and their usage.
 
 .. toctree::
    :maxdepth: 1
 
-   disk-layout
-   efi-boot
    software-update
+   software-update-server
```
Comment (dbkinder) [inline doc/architecture/architecture.rst]: doh.  @pohly and I conferred and this should really be in the howto area, so please remove the reference to software-update-server here.  Sorry.

File: doc/architecture/architecture.rst [hunk 22]
```diff
@@ -1,24 +1,254 @@
...
+The responsibility of system services is to bring the system up,
+manage connectivity and set up process' inter-process communication (IPC).
+The components Ostro OS use are commonly found in many open-source OSes: systemd,
+ConnMan, BlueZ, D-Bus and others.
+
+In addition to systemd and connectivity, Ostro OS comes with
+:ref:`software-update` technology that helps support software
+updates to deployed devices. This technology is also used to replace traditional
+package management. This means Ostro OS comes with no RPMs or DEBs support. However,
+it can still be enabled if such feature is needed in a product/device.
```
Comment (barbieri) [inline doc/architecture/architecture.rst]: as I wrote in my email to the list, to avoid getting ourselves into a corner, we should rephrase it to emphasize product updates rather than comparing to package managers. A proposal is:

"In addition to systemd and connectivity, Ostro OS comes with software-update technology tailored for IoT products. This means efficient delivery of selected and quality assured software from a single, vendor controlled, stream. The update technology copes with the fast paced IoT environment and allows live updates, reducing downtime by only restarting needed services."

File: doc/architecture/architecture.rst [hunk 23]
```diff
@@ -1,24 +1,254 @@
...
+
+Ostro OS 1.0 focuses on headless use-cases and thus no graphics and
+multimedia are enabled by default.
```
Comment (barbieri) [inline doc/architecture/architecture.rst]: move this under "Base Libraries", you can rephrase as:

"While Ostro OS 1.0 is focused on headless use-cases and no graphic and multimedia libraries are enabled, one can use the Yocto Project to incorporate them to your product."

File: doc/architecture/architecture.rst [hunk 24]
```diff
@@ -1,24 +1,254 @@
...
+
+Ostro OS Development Workflows
+==============================
+
+.. _`Yocto Project extensible SDK`: http://www.yoctoproject.org/docs/2.1/sdk-manual/sdk-manual.html#sdk-extensible
+
+Ostro OS enables many options how to do hacking and/or IoT product development.
+
+#. **Quick Prototyping**: Using the pre-built development images found in `Ostro Project download server`_ it's easy to get started with any of the :ref:`platforms`. Some of the image configurations also provide an environment to do quick prototyping with programming tools discussed in 'Programming Interfaces' above.
+#. **Image Customization**: If the pre-built development images do not provide the desired content or configuration, images can easily be customized to suit the needs. Alternatively, :ref:`software-update` could be used to extend the image content by installing additional bundles.
```
Comment (barbieri) [inline doc/architecture/architecture.rst]: Rephrase this thinking about products. There is no "if", products will customize the images in a way :-)

Remove the bundles mention, confuses the reader about the customizations

File: doc/architecture/architecture.rst (hunk 20 above)
Comment (dbkinder) [inline doc/architecture/architecture.rst]: Here's a Gustavo summary that could work here: "The application framework is simple and focused on how to start each application in its desired environment. It implements a `systemd.generator`_ that can generate a systemd service based on the application type as defined in its manifest. Currently, the supported types will provide services that will open the firewall ports, use the correct user and groups, sandbox in a container (if desired), and stitch the file system as desired by the application. This means applications may have their own file system view, with parts of the system being hidden or write-protected, and they may have a private network or even no network at all."

**Comment by mythi** [comment]
version X: updated based on comments from @barbieri and @dbkinder. Thanks!

**Comment by mythi** [comment]
rebased and removed meta-edison-bsp link. Enough reviews so merging
